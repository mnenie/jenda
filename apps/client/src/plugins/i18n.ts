import { createI18n } from 'vue-i18n'
import { enZod, ruZod, zhZod } from '@/shared/libs/vee-validate/rules'
import { customPluralRule } from '@/shared/libs/i18n/plurals'
import { getCurrentLocale } from '@/shared/libs/i18n/utils/getLocale'

interface MessageSchema {
  [key: string]: any
}

const ru_ru = import.meta.glob(
  '@/shared/libs/i18n/locales/**/ru-RU.ts',
  { eager: true },
)
const en_us = import.meta.glob(
  '@/shared/libs/i18n/locales/**/en-US.ts',
  { eager: true },
)
const zh_cn = import.meta.glob(
  '@/shared/libs/i18n/locales/**/zh-CN.ts',
  { eager: true },
)

function loadLang(modules: Record<string, any>) {
  const messages: { [key: string]: string } = {}

  Object.keys(modules).forEach((module) => {
    Object.assign(messages, { ...modules[module].default })
  })
  return messages
}

const messages = {
  'en-US': {
    ...loadLang(en_us),
    errors: {
      ...enZod,
    },
  },
  'ru-RU': {
    ...loadLang(ru_ru),
    errors: {
      ...ruZod,
    },
  },
  'zh-CN': {
    ...loadLang(zh_cn),
    errors: {
      ...zhZod,
    },
  },
}

const i18n = createI18n<[MessageSchema], 'en-US' | 'ru-RU' | 'zh-CN'>({
  legacy: false,
  locale: getCurrentLocale.value,
  globalInjection: true,
  pluralRules: {
    'ru-RU': {
      // @ts-expect-error i18n shema
      customPluralRule,
    },
  },
  messages,
})

export default i18n
