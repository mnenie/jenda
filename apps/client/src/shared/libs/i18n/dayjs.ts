import { watch } from 'vue'
import dayjs from 'dayjs'
import 'dayjs/locale/en'
import 'dayjs/locale/ru'
import 'dayjs/locale/zh-cn'
import type { I18n } from 'vue-i18n'

export default function useDayJsLocalize<
  T extends I18n<Record<string, string>, Record<string, string>>,
>(
  i18n: T,
) {
  const dayJsLocaleMap: Record<string, string> = {
    'en-US': 'en',
    'ru-RU': 'ru',
    'zh-CN': 'zh-cn',
  }

  watch(
    // @ts-ignore
    () => i18n.global!.locale.value,
    (newLocale) => {
      dayjs.locale(dayJsLocaleMap[newLocale as keyof typeof dayJsLocaleMap])
    },
    { immediate: true },
  )
}
