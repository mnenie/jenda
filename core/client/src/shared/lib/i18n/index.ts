import { useCookies } from '@vueuse/integrations/useCookies'
import { computed } from 'vue'
import { createI18n } from 'vue-i18n'
import enLocale from './locales/en-US'
import ruLocale from './locales/ru-RU'
import zhLocale from './locales/zh-CN'

type MessageSchema = typeof enLocale

const messages = {
  'en-US': {
    ...enLocale,
  },
  'ru-RU': {
    ...ruLocale,
  },
  'zh-CN': {
    ...zhLocale,
  },
}
const cookies = useCookies()

const getCurrentLocale = computed(() => {
  const cookieLanguage = cookies.get('i18n')
  if (cookieLanguage) {
    return cookieLanguage
  }
  if (navigator.language.split('-')[0] === 'ru') {
    return 'ru-RU'
  }
  return 'en-US'
})

const i18n = createI18n<[MessageSchema], 'en-US' | 'ru-RU' | 'zh-CN'>({
  legacy: false,
  locale: getCurrentLocale.value,
  globalInjection: true,
  messages,
})

export default i18n
