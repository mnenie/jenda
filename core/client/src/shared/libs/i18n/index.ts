import { computed } from 'vue'
import { createI18n } from 'vue-i18n'

import en from '../vee-validate/rules/en'
import ru from '../vee-validate/rules/ru'
import zh from '../vee-validate/rules/zh'
import enLocale from './locales/en-US'
import ruLocale from './locales/ru-RU'
import zhLocale from './locales/zh-CN'

export type MessageSchema = typeof enLocale

const messages = {
  'en-US': {
    ...enLocale,
    errors: {
      ...en,
    },
  },
  'ru-RU': {
    ...ruLocale,
    errors: {
      ...ru,
    },
  },
  'zh-CN': {
    ...zhLocale,
    errors: {
      ...zh,
    },
  },
}

const getCurrentLocale = computed(() => {
  const storageLanguage = localStorage.getItem('i18n')
  if (storageLanguage) {
    return storageLanguage
  }
  if (navigator.language.split('-')[0] === 'ru') {
    return 'ru-RU'
  }
  return 'en-US'
})

const i18n = createI18n<[MessageSchema], 'en-US' | 'ru-RU' | 'zh-CN'>({
  legacy: false,
  locale: getCurrentLocale.value,
  globalInjection: true,
  messages,
})

export default i18n
