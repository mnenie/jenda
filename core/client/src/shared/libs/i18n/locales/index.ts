export { default as enLocale } from './en-US'
export { default as ruLocale } from './ru-RU'
export { default as zhLocale } from './zh-CN'
