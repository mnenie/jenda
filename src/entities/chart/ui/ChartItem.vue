<script setup lang="ts">
import type { Chart } from '../model';

defineProps<{
  chart: Chart;
}>();
</script>

<template>
  <div :class="$style.chart_container">
    <h4 class="heading-4">{{ chart.title }}</h4>
    <p class="text-sm">
      {{ chart.description }}
    </p>
    <div :class="$style.chart">
      <component :is="chart.chart" />
    </div>
  </div>
</template>

<style module lang="scss">
.chart_container {
  display: flex;
  flex-direction: column;
  width: 100%;

  & > p {
    color: var(--zinc-500);
  }

  .chart {
    width: 100%;
    height: 100%;
    margin-top: 20px;
  }
}

:global(html.dark) {
  .chart_container {
    & > p {
      color: var(--zinc-300);
    }
  }
}
</style>
