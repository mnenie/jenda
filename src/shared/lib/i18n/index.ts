import { createI18n } from 'vue-i18n';
import { useCookies } from '@vueuse/integrations/useCookies';
import enLocale from './locales/en-US';
import ruLocale from './locales/ru-RU';
import zhLocale from './locales/zh-CN';

type MessageSchema = typeof enLocale;

const messages = {
  'en-US': {
    ...enLocale
  },
  'ru-RU': {
    ...ruLocale
  },
  'zh-CN': {
    ...zhLocale
  }
};
const cookies = useCookies();

const getCurrentLocale = () => {
  const cookieLanguage = cookies.get('i18n');
  if (cookieLanguage) {
    return cookieLanguage;
  }

  return 'en-US';
};

const i18n = createI18n<[MessageSchema], 'en-US' | 'ru-RU' | 'zh-CN'>({
  legacy: false,
  locale: getCurrentLocale(),
  messages: messages
});

export default i18n;
