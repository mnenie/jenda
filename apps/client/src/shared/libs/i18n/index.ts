import { createI18n, type I18n } from 'vue-i18n'
import { enZod, ruZod, zhZod } from '../vee-validate/rules'
import { enLocale, ruLocale, zhLocale } from './locales'
import { customPluralRule } from './plurals'
import useDayJsLocalize from './dayjs'
import { getCurrentLocale } from './utils/getLocale'

export type MessageSchema = typeof enLocale

const messages = {
  'en-US': {
    ...enLocale,
    errors: {
      ...enZod,
    },
  },
  'ru-RU': {
    ...ruLocale,
    errors: {
      ...ruZod,
    },
  },
  'zh-CN': {
    ...zhLocale,
    errors: {
      ...zhZod,
    },
  },
}

const i18n = createI18n<[MessageSchema], 'en-US' | 'ru-RU' | 'zh-CN'>({
  legacy: false,
  locale: getCurrentLocale.value,
  globalInjection: true,
  pluralRules: {
    'ru-RU': {
      // @ts-expect-error i18n shema
      customPluralRule,
    },
  },
  messages,
})

useDayJsLocalize(i18n as I18n)

export default i18n
