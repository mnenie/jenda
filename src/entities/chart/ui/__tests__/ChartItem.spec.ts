import { describe, it, expect, vi } from 'vitest';
import { shallowMount } from '@vue/test-utils';
import ChartItem from '../ChartItem.vue';
import { defineComponent, h } from 'vue';
import i18n from '@/shared/lib/i18n';

vi.mock('@vueuse/integrations/useCookies', () => {
  return {
    useCookies: () => ({
      get(key: string) {
        return key === 'i18n' ? 'en-US' : undefined;
      }
    })
  };
});

const TestChartComponent = defineComponent({
  render() {
    return h('div', 'chart');
  }
});

describe('tests for ChartItem.vue ', () => {
  const wrapper = shallowMount(ChartItem, {
    global: {
      plugins: [i18n],
      mocks: {
        t: (key: string) => {
          const translations: Record<string, string> = {
            'boards.chart.test1': 'test chart title',
            'boards.chart.test2': 'test chart description'
          };
          return translations[key];
        }
      }
    },
    props: {
      chart: {
        id: 'test-chart',
        titleKeyI18n: 'test-title',
        descriptionKeyI18n: 'test-description',
        chart: TestChartComponent
      }
    }
  });

  it('should render correctly', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
