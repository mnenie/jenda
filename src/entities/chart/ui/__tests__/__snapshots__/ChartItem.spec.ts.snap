// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tests for ChartItem.vue  > should render correctly 1`] = `
"<div class="chart_container">
  <h4 class="heading-4"></h4>
  <p class="text-sm"></p>
  <div class="chart">
    <anonymous-stub></anonymous-stub>
  </div>
</div>"
`;
